cilium: split off handling of lb svc type in datapath

borkmann opened the pull request:
See commit msgs.

borkmann committed:
bpf: split-off LoadBalancer from externalIPs svc type

Relying on externalIPs services type for LoadBalancer is not correct, since the
expectation is that E-W traffic on a LoadBalancer service must also be translated
for a Cilium-managed node's host-namespace. This is currently not the case given
we don't perform such translation on externalIPs. Thus split it off into its own
type. Semantics are: N-S via bpf_{host,xdp} similar to how we treat NodePort and
externalIP traffic today, so this stays as-is for LoadBalancer. E-W we handle it
similarly as we do in bpf_sock for ClusterIP, meaning direct match in the svc map
and unconditional translation if a service was found. This is different to NodePort
where we additionally do a wild-card lookup or externalIPs where we heavily restrict
the translation for E-W.

Related: https://github.com/cilium/cilium/issues/11724#issuecomment-646922042
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: do not remap LoadBalancer in agent code anymore

As of today, we were remapping lb.SVCTypeLoadBalancer to lb.SVCTypeExternalIPs
service in our orchestration code. Remove this implementation detail and pass
down the actual type into the BPF service map since we now have LB as its
own type.

Tested local nginx deployment w/ type LoadBalancer on metallb:

  # kubectl get svc
  NAME         TYPE           CLUSTER-IP       EXTERNAL-IP       PORT(S)        AGE
  kubernetes   ClusterIP      10.96.0.1        <none>            443/TCP        49m
  nginx        LoadBalancer   10.109.156.183   192.168.178.100   80:31260/TCP   24m

As can be seen k8s adds a ClusterIP, NodePort and LoadBalancer type:

  # ./cilium/cilium service list
  ID   Frontend               Service Type   Backend
  1    10.96.0.1:443          ClusterIP      1 => 192.168.178.29:6443
  2    10.96.0.10:9153        ClusterIP      1 => 10.217.1.44:9153
                                             2 => 10.217.1.84:9153
  3    10.96.0.10:53          ClusterIP      1 => 10.217.1.44:53
                                             2 => 10.217.1.84:53
  4    10.109.156.183:80      ClusterIP      1 => 10.217.1.43:80        <---
  5    192.168.178.29:31260   NodePort       1 => 10.217.1.43:80        <---
  6    0.0.0.0:31260          NodePort       1 => 10.217.1.43:80        <---
  7    192.168.178.100:80     LoadBalancer   1 => 10.217.1.43:80        <---

From BPF map PoV:

  # ./cilium/cilium bpf lb list
  SERVICE ADDRESS        BACKEND ADDRESS
  192.168.178.100:80     0.0.0.0:0 (7) [LoadBalancer]                   <---
                         10.217.1.43:80 (7)                             <---
  192.168.178.29:31260   0.0.0.0:0 (5) [NodePort]                       <---
                         10.217.1.43:80 (5)                             <---
  0.0.0.0:31260          10.217.1.43:80 (6)                             <---
                         0.0.0.0:0 (6) [NodePort]                       <---
  10.96.0.10:9153        10.217.1.44:9153 (2)
                         10.217.1.84:9153 (2)
                         0.0.0.0:0 (2) [ClusterIP]
  10.109.156.183:80      0.0.0.0:0 (4) [ClusterIP]                      <---
                         10.217.1.43:80 (4)                             <---
  10.96.0.1:443          0.0.0.0:0 (1) [ClusterIP]
                         192.168.178.29:6443 (1)
  10.96.0.10:53          10.217.1.44:53 (3)
                         10.217.1.84:53 (3)
                         0.0.0.0:0 (3) [ClusterIP]

Also add `cilium service update` support for aiding in manual testing.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -71,10 +72,12 @@ func CreateSvcFlag(svcLocal, sessionAffinity bool, svcTypes ...SVCType) ServiceF
-			flags |= serviceFlagHostPort
 		case SVCTypeNodePort:
 			flags |= serviceFlagNodePort
+		case SVCTypeLoadBalancer:
```
Might be worth to expand `TestCreateSvcFlag` to cover `SVCTypeLoadBalancer`?

tklauser commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -120,8 +125,8 @@ func (s ServiceFlags) SVCTrafficPolicy() SVCTrafficPolicy {
 	typeSet := false
-	sType := s & (serviceFlagExternalIPs | serviceFlagHostPort | serviceFlagNodePort)
-	for _, svcType := range []SVCType{SVCTypeExternalIPs, SVCTypeHostPort, SVCTypeNodePort} {
+	sType := s & (serviceFlagExternalIPs | serviceFlagHostPort | serviceFlagNodePort | serviceFlagLoadBalancer)
```
Same here, maybe extend `TestServiceFlags_String` to cover `SVCTypeLoadBalancer`?

tklauser reviewed (approved):
LGTM. Two small remarks regarding extending the unit tests. Probably not worth respinning just for these.

aanm reviewed (changes requested):
LGTM but needs e2e that we can run on GKE since it offers a LB

borkmann committed:
test: add svc flags unit test for LoadBalancer

Add a flag/string test each.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docs: emphasize both LoadBalancer and NodePort in XDP section

Make it more clear that this works for both and is also supported with
MetalLB for on-prem, for example.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test: add GKE tests to LoadBalancer type

Add test cases for Host/Pod connectivity on GKE for LoadBalancer.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
